### 🟦 Importing Libraries

In [1]:
import pandas as pd
import os

### 🟦 Scrutinizing & Editing Data

In [2]:
data = pd.read_csv("https://raw.githubusercontent.com/Alireza-Esp/VFCPM/refs/heads/main/Data/data.csv")

In [3]:
data

Model year   Make          Model                    Vehicle class  \
0            1995  Acura        Integra                       Subcompact   
1            1995  Acura        Integra                       Subcompact   
2            1995  Acura   Integra GS-R                       Subcompact   
3            1995  Acura         Legend                          Compact   
4            1995  Acura   Legend Coupe                          Compact   
...           ...    ...            ...                              ...   
27808        2024  Volvo  V60 CC B5 AWD             Station wagon: Small   
27809        2024  Volvo  V90 CC B6 AWD          Station wagon: Mid-size   
27810        2024  Volvo    XC40 B5 AWD     Sport utility vehicle: Small   
27811        2024  Volvo    XC60 B5 AWD     Sport utility vehicle: Small   
27812        2024  Volvo    XC90 B6 AWD  Sport utility vehicle: Standard   

       Engine size (L)  Cylinders Transmission Fuel type  City (L/100 km)  \
0                  1.8          4           A4         X             11.6   
1                  1.8          4           M5         X             11.0   
2                  1.8          4           M5         Z             10.8   
3                  3.2          6           A4         Z             14.2   
4                  3.2          6           A4         Z             14.6   
...                ...        ...          ...       ...              ...   
27808              2.0          4          AS8         Z             10.0   
27809              2.0          4          AS8         Z             10.7   
27810              2.0          4          AS8         Z              9.9   
27811              2.0          4          AS8         Z             10.6   
27812              2.0          4          AS8         Z             10.9   

       Highway (L/100 km)  Combined (L/100 km)  Combined (mpg)  \
0                     8.3                 10.1              28   
1                     8.3                  9.8              29   
2                     8.3                  9.7              29   
3                    10.5                 12.5              23   
4                    11.0                 13.0              22   
...                   ...                  ...             ...   
27808                 7.5                  8.9              32   
27809                 8.1                  9.5              30   
27810                 7.8                  9.0              31   
27811                 8.3                  9.6              29   
27812                 8.6                  9.8              29   

       CO2 emissions (g/km)  CO2 rating  Smog rating  
0                       232         NaN          NaN  
1                       225         NaN          NaN  
2                       223         NaN          NaN  
3                       288         NaN          NaN  
4                       299         NaN          NaN  
...                     ...         ...          ...  
27808                   207         5.0          5.0  
27809                   223         5.0          7.0  
27810                   210         5.0          5.0  
27811                   226         5.0          5.0  
27812                   230         5.0          7.0  

[27813 rows x 15 columns]

🔷 About 28000 sample

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27813 entries, 0 to 27812
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model year            27813 non-null  int64  
 1   Make                  27813 non-null  object 
 2   Model                 27813 non-null  object 
 3   Vehicle class         27813 non-null  object 
 4   Engine size (L)       27813 non-null  float64
 5   Cylinders             27813 non-null  int64  
 6   Transmission          27813 non-null  object 
 7   Fuel type             27813 non-null  object 
 8   City (L/100 km)       27813 non-null  float64
 9   Highway (L/100 km)    27813 non-null  float64
 10  Combined (L/100 km)   27813 non-null  float64
 11  Combined (mpg)        27813 non-null  int64  
 12  CO2 emissions (g/km)  27813 non-null  int64  
 13  CO2 rating            8832 non-null   float64
 14  Smog rating           7726 non-null   float64
dtypes: float64(6), int6

🔷 The columns "CO2 rating" and "Smog rating" have about 30% NULL values. deleting them...

In [5]:
data.drop(columns=["CO2 rating", "Smog rating"], inplace=True)

In [6]:
data

Model year   Make          Model                    Vehicle class  \
0            1995  Acura        Integra                       Subcompact   
1            1995  Acura        Integra                       Subcompact   
2            1995  Acura   Integra GS-R                       Subcompact   
3            1995  Acura         Legend                          Compact   
4            1995  Acura   Legend Coupe                          Compact   
...           ...    ...            ...                              ...   
27808        2024  Volvo  V60 CC B5 AWD             Station wagon: Small   
27809        2024  Volvo  V90 CC B6 AWD          Station wagon: Mid-size   
27810        2024  Volvo    XC40 B5 AWD     Sport utility vehicle: Small   
27811        2024  Volvo    XC60 B5 AWD     Sport utility vehicle: Small   
27812        2024  Volvo    XC90 B6 AWD  Sport utility vehicle: Standard   

       Engine size (L)  Cylinders Transmission Fuel type  City (L/100 km)  \
0                  1.8          4           A4         X             11.6   
1                  1.8          4           M5         X             11.0   
2                  1.8          4           M5         Z             10.8   
3                  3.2          6           A4         Z             14.2   
4                  3.2          6           A4         Z             14.6   
...                ...        ...          ...       ...              ...   
27808              2.0          4          AS8         Z             10.0   
27809              2.0          4          AS8         Z             10.7   
27810              2.0          4          AS8         Z              9.9   
27811              2.0          4          AS8         Z             10.6   
27812              2.0          4          AS8         Z             10.9   

       Highway (L/100 km)  Combined (L/100 km)  Combined (mpg)  \
0                     8.3                 10.1              28   
1                     8.3                  9.8              29   
2                     8.3                  9.7              29   
3                    10.5                 12.5              23   
4                    11.0                 13.0              22   
...                   ...                  ...             ...   
27808                 7.5                  8.9              32   
27809                 8.1                  9.5              30   
27810                 7.8                  9.0              31   
27811                 8.3                  9.6              29   
27812                 8.6                  9.8              29   

       CO2 emissions (g/km)  
0                       232  
1                       225  
2                       223  
3                       288  
4                       299  
...                     ...  
27808                   207  
27809                   223  
27810                   210  
27811                   226  
27812                   230  

[27813 rows x 13 columns]

In [7]:
data["Model year"].value_counts().to_dict()

{2015: 1128,
 2016: 1106,
 2009: 1096,
 2012: 1091,
 2011: 1085,
 2018: 1083,
 2013: 1082,
 2008: 1079,
 2014: 1068,
 2017: 1058,
 2019: 1056,
 2007: 1043,
 2010: 1030,
 2005: 1019,
 2022: 976,
 2020: 975,
 2021: 970,
 2006: 968,
 2004: 898,
 2023: 840,
 1995: 839,
 2003: 820,
 2024: 768,
 2002: 740,
 1996: 698,
 1999: 688,
 2001: 679,
 1997: 658,
 2000: 638,
 1998: 634}

🔷 The column "Model year" has no problem...

In [8]:
data["Make"].value_counts().to_dict()

{'Chevrolet': 2615,
 'Ford': 2255,
 'BMW': 1721,
 'GMC': 1677,
 'Mercedes-Benz': 1466,
 'Toyota': 1287,
 'Dodge': 1263,
 'Porsche': 1090,
 'Audi': 962,
 'Volkswagen': 934,
 'Nissan': 915,
 'Mazda': 780,
 'Honda': 685,
 'Hyundai': 661,
 'Jeep': 659,
 'Subaru': 626,
 'Volvo': 619,
 'Kia': 575,
 'Lexus': 537,
 'Cadillac': 511,
 'Chrysler': 504,
 'MINI': 449,
 'Jaguar': 433,
 'Pontiac': 390,
 'Buick': 339,
 'Mitsubishi': 335,
 'Acura': 319,
 'Infiniti': 313,
 'Land Rover': 265,
 'Suzuki': 260,
 'Lincoln': 252,
 'Saab': 252,
 'Saturn': 216,
 'Bentley': 171,
 'Ram': 165,
 'Maserati': 161,
 'Rolls-Royce': 133,
 'Lamborghini': 123,
 'Aston Martin': 118,
 'Ferrari': 94,
 'Plymouth': 88,
 'FIAT': 86,
 'Oldsmobile': 79,
 'Genesis': 60,
 'Alfa Romeo': 58,
 'Mercury': 57,
 'Isuzu': 47,
 'Scion': 44,
 'Eagle': 41,
 'Daewoo': 41,
 'smart': 22,
 'Hummer': 21,
 'Geo': 19,
 'Bugatti': 16,
 'SRT': 4}

🔷 SRT is subcategory of Dodge.

In [9]:
data.loc[data["Make"] == "SRT", "Make"] = "Dodge"

In [10]:
data["Model"].value_counts().to_dict()

{'Mustang': 136,
 'F-150': 107,
 'Jetta': 107,
 'F-150 4X4': 98,
 'Corolla': 88,
 'Ranger': 88,
 'Silverado': 83,
 'Sierra': 83,
 'Camaro': 81,
 'Sierra 4WD': 78,
 'Silverado 4WD': 78,
 'F-150 FFV': 76,
 'Sentra': 74,
 'Camry': 73,
 'Ram 1500': 73,
 'Accord': 72,
 'Tacoma': 71,
 'K1500 Sierra 4X4': 71,
 'A4 quattro': 71,
 'Altima': 70,
 'C1500 Sierra': 70,
 'Elantra': 69,
 'Dakota': 69,
 'Passat': 68,
 'F-150 FFV 4X4': 66,
 'Malibu': 62,
 'Canyon': 62,
 'Colorado': 62,
 'Dakota 4X4': 62,
 'Frontier': 61,
 'Sonata': 59,
 'Corvette': 57,
 'Accent': 56,
 'Legacy AWD': 56,
 'Ranger 4X4': 50,
 'Compass': 50,
 'Grand Caravan': 50,
 'A6 quattro': 50,
 'Escape': 49,
 'Tacoma 4WD': 49,
 'Civic': 48,
 'Versa': 48,
 'Lancer': 47,
 'Impala': 47,
 'Rio': 46,
 'Forester AWD': 46,
 'CTS': 46,
 'Grand Cherokee 4X4': 46,
 'Impreza AWD': 46,
 'Tundra': 46,
 'Mazda6': 45,
 'Forte': 45,
 'Fusion': 43,
 'Boxster': 43,
 'Sunfire': 43,
 '911 Carrera': 43,
 'Cavalier': 42,
 'Tacoma 4X4': 42,
 'Golf': 42,
 'Pa

🔷 The "Model" column should naturally be removed. Because the purpose of this model is to predict fuel consumption and CO2 emissions. These indicators are usually calculated, estimated and provided by the manufacturer for each vehicle based on its technical characteristics. As a result, predicting them is meaningless.

In [11]:
data.drop(columns=["Model"], inplace=True)

In [12]:
data

Model year   Make                    Vehicle class  Engine size (L)  \
0            1995  Acura                       Subcompact              1.8   
1            1995  Acura                       Subcompact              1.8   
2            1995  Acura                       Subcompact              1.8   
3            1995  Acura                          Compact              3.2   
4            1995  Acura                          Compact              3.2   
...           ...    ...                              ...              ...   
27808        2024  Volvo             Station wagon: Small              2.0   
27809        2024  Volvo          Station wagon: Mid-size              2.0   
27810        2024  Volvo     Sport utility vehicle: Small              2.0   
27811        2024  Volvo     Sport utility vehicle: Small              2.0   
27812        2024  Volvo  Sport utility vehicle: Standard              2.0   

       Cylinders Transmission Fuel type  City (L/100 km)  Highway (L/100 km)  \
0              4           A4         X             11.6                 8.3   
1              4           M5         X             11.0                 8.3   
2              4           M5         Z             10.8                 8.3   
3              6           A4         Z             14.2                10.5   
4              6           A4         Z             14.6                11.0   
...          ...          ...       ...              ...                 ...   
27808          4          AS8         Z             10.0                 7.5   
27809          4          AS8         Z             10.7                 8.1   
27810          4          AS8         Z              9.9                 7.8   
27811          4          AS8         Z             10.6                 8.3   
27812          4          AS8         Z             10.9                 8.6   

       Combined (L/100 km)  Combined (mpg)  CO2 emissions (g/km)  
0                     10.1              28                   232  
1                      9.8              29                   225  
2                      9.7              29                   223  
3                     12.5              23                   288  
4                     13.0              22                   299  
...                    ...             ...                   ...  
27808                  8.9              32                   207  
27809                  9.5              30                   223  
27810                  9.0              31                   210  
27811                  9.6              29                   226  
27812                  9.8              29                   230  

[27813 rows x 12 columns]

In [13]:
data["Vehicle class"].value_counts().to_dict()

{'Compact': 3837,
 'Mid-size': 3593,
 'Sport utility vehicle': 3044,
 'Pickup truck: Standard': 2961,
 'Subcompact': 2658,
 'Sport utility vehicle: Small': 2158,
 'Full-size': 1761,
 'Two-seater': 1510,
 'Sport utility vehicle: Standard': 1430,
 'Minicompact': 1136,
 'Station wagon: Small': 991,
 'Pickup truck: Small': 614,
 'Minivan': 610,
 'Station wagon: Mid-size': 524,
 'Van: Cargo': 469,
 'Van: Passenger': 398,
 'Special purpose vehicle': 119}

In [14]:
sorted(data["Vehicle class"].value_counts().to_dict())

['Compact',
 'Full-size',
 'Mid-size',
 'Minicompact',
 'Minivan',
 'Pickup truck: Small',
 'Pickup truck: Standard',
 'Special purpose vehicle',
 'Sport utility vehicle',
 'Sport utility vehicle: Small',
 'Sport utility vehicle: Standard',
 'Station wagon: Mid-size',
 'Station wagon: Small',
 'Subcompact',
 'Two-seater',
 'Van: Cargo',
 'Van: Passenger']

🔷 For simplifying the model, we change the values of "Vehicle class" in this way:

In [15]:
data.loc[data["Vehicle class"].isin(["Pickup truck: Small","Pickup truck: Standard"]), "Vehicle class" ] = "Pickup truck"

data.loc[data["Vehicle class"].isin(["Sport utility vehicle: Small","Sport utility vehicle: Standard"]), "Vehicle class" ] = "Sport utility vehicle"

data.loc[data["Vehicle class"].isin(["Station wagon: Mid-size","Station wagon: Small"]), "Vehicle class" ] = "Station wagon"

data.loc[data["Vehicle class"].isin(["Van: Cargo","Van: Passenger"]), "Vehicle class" ] = "Van"

In [16]:
data

Model year   Make          Vehicle class  Engine size (L)  Cylinders  \
0            1995  Acura             Subcompact              1.8          4   
1            1995  Acura             Subcompact              1.8          4   
2            1995  Acura             Subcompact              1.8          4   
3            1995  Acura                Compact              3.2          6   
4            1995  Acura                Compact              3.2          6   
...           ...    ...                    ...              ...        ...   
27808        2024  Volvo          Station wagon              2.0          4   
27809        2024  Volvo          Station wagon              2.0          4   
27810        2024  Volvo  Sport utility vehicle              2.0          4   
27811        2024  Volvo  Sport utility vehicle              2.0          4   
27812        2024  Volvo  Sport utility vehicle              2.0          4   

      Transmission Fuel type  City (L/100 km)  Highway (L/100 km)  \
0               A4         X             11.6                 8.3   
1               M5         X             11.0                 8.3   
2               M5         Z             10.8                 8.3   
3               A4         Z             14.2                10.5   
4               A4         Z             14.6                11.0   
...            ...       ...              ...                 ...   
27808          AS8         Z             10.0                 7.5   
27809          AS8         Z             10.7                 8.1   
27810          AS8         Z              9.9                 7.8   
27811          AS8         Z             10.6                 8.3   
27812          AS8         Z             10.9                 8.6   

       Combined (L/100 km)  Combined (mpg)  CO2 emissions (g/km)  
0                     10.1              28                   232  
1                      9.8              29                   225  
2                      9.7              29                   223  
3                     12.5              23                   288  
4                     13.0              22                   299  
...                    ...             ...                   ...  
27808                  8.9              32                   207  
27809                  9.5              30                   223  
27810                  9.0              31                   210  
27811                  9.6              29                   226  
27812                  9.8              29                   230  

[27813 rows x 12 columns]

In [17]:
data["Vehicle class"].value_counts().to_dict()

{'Sport utility vehicle': 6632,
 'Compact': 3837,
 'Mid-size': 3593,
 'Pickup truck': 3575,
 'Subcompact': 2658,
 'Full-size': 1761,
 'Station wagon': 1515,
 'Two-seater': 1510,
 'Minicompact': 1136,
 'Van': 867,
 'Minivan': 610,
 'Special purpose vehicle': 119}

In [18]:
sorted(data["Vehicle class"].value_counts().to_dict())

['Compact',
 'Full-size',
 'Mid-size',
 'Minicompact',
 'Minivan',
 'Pickup truck',
 'Special purpose vehicle',
 'Sport utility vehicle',
 'Station wagon',
 'Subcompact',
 'Two-seater',
 'Van']

Sedans: Minicompact, Subcompact, Compact, Mid-size, Full-size

Station Wagons: Station wagon

Trucks: Pickup truck, Van, Minivan, Sport utility vehicle, Special purpose vehicle

Two Seaters: Two-seater

In [19]:
data[["Engine size (L)"]].describe()

Engine size (L)
count     27813.000000
mean          3.340237
std           1.341174
min           0.800000
25%           2.200000
50%           3.000000
75%           4.200000
max           8.400000

🔷 The column "Engine size" has no problem...

In [20]:
data["Cylinders"].value_counts().to_dict()

{4: 10051, 6: 9907, 8: 6308, 5: 591, 12: 523, 3: 216, 10: 184, 2: 17, 16: 16}

In [21]:
data[["Cylinders"]].describe()

Cylinders
count  27813.000000
mean       5.828893
std        1.806877
min        2.000000
25%        4.000000
50%        6.000000
75%        8.000000
max       16.000000

🔷 The column "Cylinders" has no problem...

In [22]:
data["Transmission"].value_counts().to_dict()

{'A4': 5489,
 'M5': 3222,
 'AS6': 2911,
 'M6': 2799,
 'AS8': 2193,
 'A6': 1975,
 'A5': 1604,
 'A8': 989,
 'AV': 895,
 'AS5': 874,
 'AM7': 794,
 'A9': 650,
 'AS7': 621,
 'AS10': 506,
 'A7': 281,
 'A10': 279,
 'AS4': 258,
 'AM6': 232,
 'AM8': 232,
 'A3': 187,
 'AV6': 159,
 'AS9': 158,
 'AV7': 157,
 'M7': 151,
 'AV8': 124,
 'AV10': 39,
 'AV1': 19,
 'AM9': 6,
 'AM5': 6,
 'M4': 3}

🔷 In 895 samples, "AV" is setted for "Transmission" column. should be removed.

In [23]:
data.drop(index=list(data.loc[data["Transmission"] == "AV"].index), inplace=True)
data.reset_index(inplace=True)
data.drop(columns=["index"], inplace=True)

In [24]:
data

Model year   Make          Vehicle class  Engine size (L)  Cylinders  \
0            1995  Acura             Subcompact              1.8          4   
1            1995  Acura             Subcompact              1.8          4   
2            1995  Acura             Subcompact              1.8          4   
3            1995  Acura                Compact              3.2          6   
4            1995  Acura                Compact              3.2          6   
...           ...    ...                    ...              ...        ...   
26913        2024  Volvo          Station wagon              2.0          4   
26914        2024  Volvo          Station wagon              2.0          4   
26915        2024  Volvo  Sport utility vehicle              2.0          4   
26916        2024  Volvo  Sport utility vehicle              2.0          4   
26917        2024  Volvo  Sport utility vehicle              2.0          4   

      Transmission Fuel type  City (L/100 km)  Highway (L/100 km)  \
0               A4         X             11.6                 8.3   
1               M5         X             11.0                 8.3   
2               M5         Z             10.8                 8.3   
3               A4         Z             14.2                10.5   
4               A4         Z             14.6                11.0   
...            ...       ...              ...                 ...   
26913          AS8         Z             10.0                 7.5   
26914          AS8         Z             10.7                 8.1   
26915          AS8         Z              9.9                 7.8   
26916          AS8         Z             10.6                 8.3   
26917          AS8         Z             10.9                 8.6   

       Combined (L/100 km)  Combined (mpg)  CO2 emissions (g/km)  
0                     10.1              28                   232  
1                      9.8              29                   225  
2                      9.7              29                   223  
3                     12.5              23                   288  
4                     13.0              22                   299  
...                    ...             ...                   ...  
26913                  8.9              32                   207  
26914                  9.5              30                   223  
26915                  9.0              31                   210  
26916                  9.6              29                   226  
26917                  9.8              29                   230  

[26918 rows x 12 columns]

🔷 Separating transmission type and number of gears.

In [25]:
Number_Of_Gears = []
NOG = ""

for index,value in zip(list(data["Transmission"].keys()), list(data["Transmission"].values)):
    
    for i in value:
        if i.isnumeric() == True:
            NOG = NOG + i
            
    Number_Of_Gears.append(int(NOG))
    data.loc[index, "Transmission"] = value[:-len(NOG)]
    NOG = ""
    
data.insert(6, "Number of gears", Number_Of_Gears)

In [26]:
data

Model year   Make          Vehicle class  Engine size (L)  Cylinders  \
0            1995  Acura             Subcompact              1.8          4   
1            1995  Acura             Subcompact              1.8          4   
2            1995  Acura             Subcompact              1.8          4   
3            1995  Acura                Compact              3.2          6   
4            1995  Acura                Compact              3.2          6   
...           ...    ...                    ...              ...        ...   
26913        2024  Volvo          Station wagon              2.0          4   
26914        2024  Volvo          Station wagon              2.0          4   
26915        2024  Volvo  Sport utility vehicle              2.0          4   
26916        2024  Volvo  Sport utility vehicle              2.0          4   
26917        2024  Volvo  Sport utility vehicle              2.0          4   

      Transmission  Number of gears Fuel type  City (L/100 km)  \
0                A                4         X             11.6   
1                M                5         X             11.0   
2                M                5         Z             10.8   
3                A                4         Z             14.2   
4                A                4         Z             14.6   
...            ...              ...       ...              ...   
26913           AS                8         Z             10.0   
26914           AS                8         Z             10.7   
26915           AS                8         Z              9.9   
26916           AS                8         Z             10.6   
26917           AS                8         Z             10.9   

       Highway (L/100 km)  Combined (L/100 km)  Combined (mpg)  \
0                     8.3                 10.1              28   
1                     8.3                  9.8              29   
2                     8.3                  9.7              29   
3                    10.5                 12.5              23   
4                    11.0                 13.0              22   
...                   ...                  ...             ...   
26913                 7.5                  8.9              32   
26914                 8.1                  9.5              30   
26915                 7.8                  9.0              31   
26916                 8.3                  9.6              29   
26917                 8.6                  9.8              29   

       CO2 emissions (g/km)  
0                       232  
1                       225  
2                       223  
3                       288  
4                       299  
...                     ...  
26913                   207  
26914                   223  
26915                   210  
26916                   226  
26917                   230  

[26918 rows x 13 columns]

In [27]:
data["Fuel type"].value_counts()

Fuel type
X    14381
Z    10905
E     1098
D      495
N       39
Name: count, dtype: int64

🔷 According to the dataset description, we change the "fuel type" column.

In [28]:
data.loc[data["Fuel type"] == "X", "Fuel type"] = "Regular gasoline"

data.loc[data["Fuel type"] == "Z", "Fuel type"] = "Premium gasoline"

data.loc[data["Fuel type"] == "D", "Fuel type"] = "Diesel"

data.loc[data["Fuel type"] == "E", "Fuel type"] = "Ethanol (E85)"

data.loc[data["Fuel type"] == "N", "Fuel type"] = "Natural Gas"

In [29]:
data

Model year   Make          Vehicle class  Engine size (L)  Cylinders  \
0            1995  Acura             Subcompact              1.8          4   
1            1995  Acura             Subcompact              1.8          4   
2            1995  Acura             Subcompact              1.8          4   
3            1995  Acura                Compact              3.2          6   
4            1995  Acura                Compact              3.2          6   
...           ...    ...                    ...              ...        ...   
26913        2024  Volvo          Station wagon              2.0          4   
26914        2024  Volvo          Station wagon              2.0          4   
26915        2024  Volvo  Sport utility vehicle              2.0          4   
26916        2024  Volvo  Sport utility vehicle              2.0          4   
26917        2024  Volvo  Sport utility vehicle              2.0          4   

      Transmission  Number of gears         Fuel type  City (L/100 km)  \
0                A                4  Regular gasoline             11.6   
1                M                5  Regular gasoline             11.0   
2                M                5  Premium gasoline             10.8   
3                A                4  Premium gasoline             14.2   
4                A                4  Premium gasoline             14.6   
...            ...              ...               ...              ...   
26913           AS                8  Premium gasoline             10.0   
26914           AS                8  Premium gasoline             10.7   
26915           AS                8  Premium gasoline              9.9   
26916           AS                8  Premium gasoline             10.6   
26917           AS                8  Premium gasoline             10.9   

       Highway (L/100 km)  Combined (L/100 km)  Combined (mpg)  \
0                     8.3                 10.1              28   
1                     8.3                  9.8              29   
2                     8.3                  9.7              29   
3                    10.5                 12.5              23   
4                    11.0                 13.0              22   
...                   ...                  ...             ...   
26913                 7.5                  8.9              32   
26914                 8.1                  9.5              30   
26915                 7.8                  9.0              31   
26916                 8.3                  9.6              29   
26917                 8.6                  9.8              29   

       CO2 emissions (g/km)  
0                       232  
1                       225  
2                       223  
3                       288  
4                       299  
...                     ...  
26913                   207  
26914                   223  
26915                   210  
26916                   226  
26917                   230  

[26918 rows x 13 columns]

In [30]:
data["City (L/100 km)"].describe()

count    26918.000000
mean        13.967973
std          3.664321
min          4.000000
25%         11.400000
50%         13.500000
75%         16.000000
max         33.300000
Name: City (L/100 km), dtype: float64

🔷 The column "City (L/100 km)" has no problem...

In [31]:
data["Highway (L/100 km)"].describe()

count    26918.000000
mean        10.079178
std          2.567706
min          3.900000
25%          8.300000
50%          9.600000
75%         11.400000
max         22.100000
Name: Highway (L/100 km), dtype: float64

🔷 The column "Highway (L/100 km)" has no problem...

In [32]:
data["Combined (L/100 km)"].describe()

count    26918.000000
mean        12.218062
std          3.133751
min          4.000000
25%         10.000000
50%         11.700000
75%         13.900000
max         27.500000
Name: Combined (L/100 km), dtype: float64

🔷 The column "Combined (L/100 km)" has no problem...

🔷 The column "Combined (mpg)" is no needed. should be removed.

In [33]:
data.drop(columns=["Combined (mpg)"], inplace=True)

In [34]:
data

Model year   Make          Vehicle class  Engine size (L)  Cylinders  \
0            1995  Acura             Subcompact              1.8          4   
1            1995  Acura             Subcompact              1.8          4   
2            1995  Acura             Subcompact              1.8          4   
3            1995  Acura                Compact              3.2          6   
4            1995  Acura                Compact              3.2          6   
...           ...    ...                    ...              ...        ...   
26913        2024  Volvo          Station wagon              2.0          4   
26914        2024  Volvo          Station wagon              2.0          4   
26915        2024  Volvo  Sport utility vehicle              2.0          4   
26916        2024  Volvo  Sport utility vehicle              2.0          4   
26917        2024  Volvo  Sport utility vehicle              2.0          4   

      Transmission  Number of gears         Fuel type  City (L/100 km)  \
0                A                4  Regular gasoline             11.6   
1                M                5  Regular gasoline             11.0   
2                M                5  Premium gasoline             10.8   
3                A                4  Premium gasoline             14.2   
4                A                4  Premium gasoline             14.6   
...            ...              ...               ...              ...   
26913           AS                8  Premium gasoline             10.0   
26914           AS                8  Premium gasoline             10.7   
26915           AS                8  Premium gasoline              9.9   
26916           AS                8  Premium gasoline             10.6   
26917           AS                8  Premium gasoline             10.9   

       Highway (L/100 km)  Combined (L/100 km)  CO2 emissions (g/km)  
0                     8.3                 10.1                   232  
1                     8.3                  9.8                   225  
2                     8.3                  9.7                   223  
3                    10.5                 12.5                   288  
4                    11.0                 13.0                   299  
...                   ...                  ...                   ...  
26913                 7.5                  8.9                   207  
26914                 8.1                  9.5                   223  
26915                 7.8                  9.0                   210  
26916                 8.3                  9.6                   226  
26917                 8.6                  9.8                   230  

[26918 rows x 12 columns]

In [35]:
data["CO2 emissions (g/km)"].describe()

count    26918.000000
mean       277.903819
std         64.611739
min         94.000000
25%        232.000000
50%        270.000000
75%        317.000000
max        633.000000
Name: CO2 emissions (g/km), dtype: float64

🔷 The column "CO2 emissions (g/km)" has no problem...

### 🟦 Saving Data

In [36]:
data.to_csv(path_or_buf=os.path.basename("\Data\data-cleaned.csv"), sep=",", index=False, encoding="utf-8")